clang/docs/BoundsSafety.rst

```diff
@@ -0,0 +1,362 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A pointer type used in a C-style cast (e.g., ``(int *)src``) inherits the same pointer attribute in the type of src. For instance, if the type of src is ``T *__single`` (with ``T`` being an arbitrary C type), ``(int *)src`` will be ``int *__single``. The reasoning behind this behavior is so that a C-style cast doesn’t introduce any unexpected side effects caused by an implicit cast of bounds attribute.
+
+Pointer casts can have explicit bounds annotations. For instance, ``(int *__bidi_indexable)src`` casts to ``int *__bidi_indexable`` as long as src has a bounds annotation that can implicitly convert to ``__bidi_indexable``. If ``src`` has type ``int *__single``, it can implicitly convert to ``int *__bidi_indexable`` which then will have the upper bound pointing to one past the first element. However, if src has type ``int *__unsafe_indexable``, the explicit cast ``(int *__bidi_indexable)src`` will cause an error because ``__unsafe_indexable`` cannot cast to ``__bidi_indexable`` as ``__unsafe_indexable`` doesn’t have bounds information. `Cast rules`_ describes in more detail what kinds of casts are allowed between pointers with different bounds annotations.
+
+Default pointer types in typedef
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Pointer types in ``typedef``s do not have implicit default bounds annotations. Instead, the bounds annotation is determined when the ``typedef`` is used. The following example shows that no pointer annotation is specified in the ``typedef pint_t`` while each instance of ``typedef``'ed pointer gets its bounds annotation based on the context in which the type is used.
+
+.. code-block:: c
+
+   typedef int * pint_t; // int *
+
+   pint_t glob; // int *__single glob;
+
+   void foo(void) {
+   pint_t local; // int *__bidi_indexable local;
+   }
+
+Pointer types in a ``typedef`` can still have explicit annotations, e.g., ``typedef int *__single``, in which case the bounds annotation ``__single`` will apply to every use of the ``typedef``.
+
+Array to pointer promotion
+--------------------------
+
+In C, when an array is referenced, it is automatically promoted (or “decayed”) to a pointer to its first element (e.g., ``&arr[0]``). Similarly, in ``-fbounds-safety``, arrays are also promoted to pointers, but with the addition of an implicit bounds annotation. Arrays on function parameters are promoted to corresponding ``__counted_by`` pointers. Consequently, incomplete arrays (or arrays without size) will cause a compiler error unless it has ``__counted_by`` annotation in its bracket. All other arrays are promoted to ``__bidi_indexable`` pointers, with the equivalent of ``&arr[0]`` serving as the lower bound and ``&arr[array_size]`` (or one past the last element) serving as the upper bound. This way, all array accesses are subject to bounds checking, just as their corresponding pointers are.
+
+Maintaining correctness of bounds annotations
```

Comment:
One thing that's not clear is how this interacts with optimizations around undefined behavior; can the optimizer remove these checks? Can it do so if it thinks the situation can't happen because it would be UB if it did and the optimizer assumes no UB?

Reply:
The optimizer may be able to remove checks if the pointer is derived from `__counted_by` and the size is known at compile time. I added some details to explain at the end of "Maintaining correctness of bounds annotations".

Reply:
Thanks! The root of my concern was time travel optimizations, but I'm not certain we can run into such a problem with the way you've designed things. e.g., (using null pointers instead of bounds)
```
void func(int *ip) {
  int i = *ip;

  if (ip) {
    foo(i);
  }
}
```
where the optimizer sees the dereference and decides it can remove the `if` statement as a result. I think you can't run into a similar problem though because of the instrumented checks you add. e.g.,
```
void func(int n, int *ip __counted_by(n)) {
  int *end = ip + 100; // Oops, should be + n instead!

  ...
}
```
I was concerned the optimizer may decide this means ip + 100 must be a valid range because otherwise it would be UB. But I think the instrumentation code will add actual bounds checking there and so the optimizer won't make that assumption. Does that sound about right to you?

Reply:
Thanks, that's great!